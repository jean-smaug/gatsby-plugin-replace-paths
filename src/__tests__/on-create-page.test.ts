import onCreatePage from "../on-create-page";

const createGatsby = ({
  createPage = jest.fn(),
  deletePage = jest.fn(),
  page = {},
  panic = jest.fn(() => {
    expect("I have valid options.").toBe(true);
  }),
} = {}) => ({
  boundActionCreators: {
    createPage,
    deletePage,
  },
  page,
  reporter: {
    panic,
  },
});

describe("onCreatePage", () => {
  it("creates a new page/delete old page when replacement is made", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something/" };
    const gatsby = createGatsby({ createPage, deletePage, page });
    const options = [{ pattern: "some", replacement: "" }];
    const actual = await onCreatePage(gatsby, options);

    expect(createPage).toHaveBeenCalled();
    expect(deletePage).toHaveBeenCalled();
    expect(createPage.mock.calls[0][0]).toMatchSnapshot("createPage");
    expect(deletePage.mock.calls[0][0]).toMatchSnapshot("deletePage");
  });

  it("supports regex for pattern option", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something-sweet/" };
    const gatsby = createGatsby({ createPage, deletePage, page });
    const options = [{ pattern: /e/g, replacement: undefined }];
    const actual = await onCreatePage(gatsby, options);

    expect(createPage).toHaveBeenCalled();
    expect(deletePage).toHaveBeenCalled();
    expect(createPage.mock.calls[0][0]).toMatchSnapshot("createPage");
    expect(deletePage.mock.calls[0][0]).toMatchSnapshot("deletePage");
  });

  it("supports function for replacement option", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something-sweet/" };
    const gatsby = createGatsby({ createPage, deletePage, page });
    const replacement = (_, match) => match.toUpperCase();
    const options = [{ pattern: /(e)/g, replacement }];
    const actual = await onCreatePage(gatsby, options);

    expect(createPage).toHaveBeenCalled();
    expect(deletePage).toHaveBeenCalled();
    expect(createPage.mock.calls[0][0]).toMatchSnapshot("createPage");
    expect(deletePage.mock.calls[0][0]).toMatchSnapshot("deletePage");
  });

  it("calls reporter.panic when pattern option is missing", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something-sweet/" };
    const panic = jest.fn((error) => error);
    const gatsby = createGatsby({ createPage, deletePage, page, panic });
    const options = [{ pattern: undefined, replacement: "" }];
    const actual = await onCreatePage(gatsby, options);

    expect(panic.mock.calls[0][0]).toMatchSnapshot();
  });

  it("calls reporter.panic when replacement option is missing", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something-sweet/" };
    const panic = jest.fn((error) => error);
    const gatsby = createGatsby({ createPage, deletePage, page, panic });
    const options = [{ pattern: /e/, replacement: null }];
    const actual = await onCreatePage(gatsby, options);

    expect(panic.mock.calls[0][0]).toMatchSnapshot();
  });

  it("does nothing when no match found", async () => {
    const createPage = jest.fn();
    const deletePage = jest.fn();
    const page = { path: "/something-sweet/" };
    const gatsby = createGatsby({ createPage, deletePage, page });
    const options = [{ pattern: /\d/, replacement: "" }];
    const actual = await onCreatePage(gatsby, options);

    expect(createPage).not.toHaveBeenCalled();
    expect(deletePage).not.toHaveBeenCalled();
  });
});
