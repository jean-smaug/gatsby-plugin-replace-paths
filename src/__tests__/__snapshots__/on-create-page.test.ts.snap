// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onCreatePage calls reporter.panic when pattern option is missing 1`] = `[Error: Must supply \`pattern\` option as a regex or a string.]`;

exports[`onCreatePage calls reporter.panic when replacement option is missing 1`] = `[Error: Must supply \`replacement\` option that is a function or a string.]`;

exports[`onCreatePage creates a new page/delete old page when replacement is made: createPage 1`] = `
Object {
  "path": "/thing/",
}
`;

exports[`onCreatePage creates a new page/delete old page when replacement is made: deletePage 1`] = `
Object {
  "path": "/something/",
}
`;

exports[`onCreatePage supports function for replacement option: createPage 1`] = `
Object {
  "path": "/somEthing-swEEt/",
}
`;

exports[`onCreatePage supports function for replacement option: deletePage 1`] = `
Object {
  "path": "/something-sweet/",
}
`;

exports[`onCreatePage supports regex for pattern option: createPage 1`] = `
Object {
  "path": "/somthing-swt/",
}
`;

exports[`onCreatePage supports regex for pattern option: deletePage 1`] = `
Object {
  "path": "/something-sweet/",
}
`;
